algebra2/basic_algebra_project/basic_algebra_2_pack.hpp

```diff
@@ -0,0 +1,670 @@
+#pragma once
+#include "gaussian_elimination.hpp"
+
+namespace algebra {
+
+/*
+  description:
+    Alias type for matrix with data vector and matrix view
+*/
+template <typename T, typename Layout>
+using matrix = std::pair<std::vector<T>, ranges::matrix_view<T, Layout>>;
+
+/*
+  description:
+    Alias type for homogeneous solution
+*/
+template <typename T>
+using homogeneous_solution_t = std::optional<matrix<T, std::layout_left>>;
+
+/*
+  description:
+    Conversion type to fraction
+*/
+template <typename T>
+using to_fraction_type =
+    std::conditional_t<algorithms::gaussian_elimination::is_fraction_v<T>, T,
+                       algorithms::gaussian_elimination::fraction<T>>;
+
+/*
+  description:
+    The function is used to find the first non-zero element in a specific row of
+  the matrix
+*/
+template <typename T, typename LP>
+auto first_non_zero(ranges::matrix_view<T, LP> m, std::size_t row_index)
+    -> std::size_t;
+
+/*
+  description:
+    The function checks whether there is a row in the given matrix that can be
+  considered contradictory.
+*/
+template <typename T, typename LP>
+auto is_contradictory(ranges::matrix_view<T, LP> combined) -> bool;
+
+/*
+  description:
+    The function it is used to find the indexes of the first non-zero elements
+  (pivots) in each row of the matrix and return them as a vector.
+*/
+template <typename T, typename LP>
+auto pivots(ranges::matrix_view<T, LP> combined) -> std::vector<std::size_t>;
```

Comment:
such methods can be safely moved to matrix.cxx file (it depends only on matrix structure)